### DISCUSSION TITLE
postgresql master/slave switch

### DISCUSSION THREAD

**Discussion by liukaiyuliukaiyu** [Kong/kong#9639] [Help] [answered]
I wonder, how does the latest version kong do master/slave switching when the database master is down

**Comment by kikito** [comment, marked as answer]
Kong does not have direct support for "master/slave hotswitching".

It does have read-only replicas support in Postgres. This means that Kong will use the replica for reading, and only use use the main db on writes. With this setup, if the main db goes down, then writes will fail, but the reads will still be able to work.

You may read more about this in https://docs.konghq.com/gateway/latest/reference/configuration/#datastore-section

**Reply by liukaiyuliukaiyu** [reply]
Your reply is greatly appreciated，thanks，I have two more questions
1, Read/write separation can ensure that the read db can work when the write db is down. Then, if the write db is down, kong needs to switch to a new write db instance, or the read db is down, kong also needs to switch to a new read db instance. I want to know how to do the switch, can we not restart kong to switch?
2, At present, the version of our kong is 0.10.3, which is old, and there is only one database to support read and write. If a proxy layer is added in front of the postgresql database, kong pg_host is configured as the VIP proxy address. If the database instance under VIP is changed, do we need to restart kong? The pg_host configuration is fixed as the VIP proxy address

**Reply by liukaiyuliukaiyu** [reply]
@kikito @hbagdi help,I look forward to and appreciate your answering the above two questions again，thanks！

**Comment by hbagdi** [comment]
An alternate approach here is to use DNS to solve the problem (we do this internally).
That requires two records: "rw.db.example.com" and "ro.db.example.com". THe "rw" endpoint points to the leader instance and the "to" endpoint points to the follower. On a failure, the DNS records should be updated which will then lead to Kong re-establishing connections (after some retries) and all will be good.
